### ISSUE TITLE
ImageDownloader and ImageProcessor issue

### ISSUE THREAD

**Issue by Mathieu-o** [onevcat/Kingfisher#457]
Hi!

I believe I've found a bug in the ImageDownloader code. Here is how it goes:
There is this optimization where if you try to download N times from the same URL you'll just do one URLSessionDataTask and call N times the callback at the end of the download.
That's all good and fine. But what we do and that doesn't work is that we launch two (or more) requests for the same URL but we use different ImageProcessor on each requests.
Because the code only stores one set of options (the first) across the N calls every callback end up with the same Image which is wrong.
Here is some (pseudo)code to illustrate :
``` swift
img1.kf.setImage(with: "http://same.url/", placeholder: nil, options: [.transition(.fade(0.3)), .processor(kRoundedProcessor)])
img2.kf.setImage(with: "http://same.url/", placeholder: nil, options: [.transition(.fade(0.3))])
```

When doing this, img2 will display the same result as img1, which is a rounded picture! 

I came up with a solution (using an array of options) which is used by our team in our product at the moment. I didn't want to make a PR because the change is quite invasive and wanted to have your opinion on it before. So here it is : 
https://gist.github.com/Mathieu-o/fea066ccf31438980da459f7bbd04a4a

Thoughts? 

Thanks for you work!

ps: Tests are all green with the patch. I case join a test case to demonstrate the issue in the PR if need be.

**Comment by onevcat** [comment]
@Mathieu-o Oops. You are definitely correct on this! Thanks for pointing this out.

I will take a look at your patch later, and decide a good way for it!

**Comment by onevcat** [comment]
Just had a look at your implementation, and it seems quite good to me!

But I guess we could refactor that a bit to make things more clear. I will give it a fix soon (maybe tomorrow) and release an update for it.

Thank you again for reporting this. Great work 👍  :)

**Comment by Mathieu-o** [comment]
Yeah well tbh I did a PoC style patch. I think grouping the callbacks and option in a tuple might be a good idea in order to avoid dealing with double indexes. And the whole repetition of what callback() is doing in processImage() is also ugly. I wanted to introduce a callback(atIndex:) variant but that didn't turn to be good.

**Comment by onevcat** [comment]
Should be fixed in #460 

I will release an update for it later today.

**Comment by onevcat** [comment]
Version 3.1.1 was released for this.
